In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/US_flights_2023.csv
/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/airports_geolocation.csv
/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/weather_meteo_by_airport.csv
/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/Cancelled_Diverted_2023.csv
/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/maj us flight - january 2024.csv


# ✈️ US Aviation Delay Analysis 2023 - Training Path

A progressive, hands-on learning journey through real-world aviation data analysis using pandas.

---

## 🎮 How This Works

Each task is designed as a **skill-building exercise** with:
- Clear learning objectives
- Step-by-step instructions
- Expected outcomes
- XP points based on difficulty

**Total XP Available:** 1,850 XP

---

## 📊 Task Overview by Category

| Category | Tasks | Total XP |
|----------|-------|----------|
| Data Loading & Inspection | 1-3 | 250 XP |
| Data Cleaning | 4-6 | 300 XP |
| Indexing & Time Operations | 7-8 | 200 XP |
| Merging & Joining | 9-12 | 450 XP |
| Feature Engineering | 13-14 | 200 XP |
| Aggregation & Analysis | 15-17 | 300 XP |
| Advanced Analytics | 18-20 | 400 XP |

---

## 🚀 Learning Path

### **LEVEL 1: Data Foundation** (250 XP)

---

### Task 1: Load and Inspect Flight Data
**Goal:** Learn to load CSV files and perform initial data exploration.

**Instructions (50 XP):**

1. Import pandas and load the main flights CSV file
2. Display the first 10 rows using `.head(10)`
3. Check the shape of the DataFrame using `.shape`
4. Display column names and data types using `.info()`
5. Generate summary statistics using `.describe()`

**Expected Outcome:** 
- Understanding of dataset structure
- Number of flights recorded
- Column types (numeric, object, datetime)
- Basic statistics (mean delays, flight counts)

**Validation Checks:**
- DataFrame loaded successfully
- Shape shows (n_rows, n_columns)
- All columns displayed with correct types

---

In [2]:
# CREATE YOUR CODE HERE

us_flights = pd.read_csv('/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/US_flights_2023.csv')
geo_airports = pd.read_csv('/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/airports_geolocation.csv')
maj_us_flight = pd.read_csv('/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/maj us flight - january 2024.csv')
weather_by_airport = pd.read_csv('/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/weather_meteo_by_airport.csv')
cancelled_diverted = pd.read_csv('/kaggle/input/2023-us-civil-flights-delay-meteo-and-aircraft/Cancelled_Diverted_2023.csv')

In [3]:
# US FLIGHTS
us_flights.head(10)

FlightDate  Day_Of_Week       Airline Tail_Number Dep_Airport  \
0  2023-01-02            1  Endeavor Air      N605LR         BDL   
1  2023-01-03            2  Endeavor Air      N605LR         BDL   
2  2023-01-04            3  Endeavor Air      N331PQ         BDL   
3  2023-01-05            4  Endeavor Air      N906XJ         BDL   
4  2023-01-06            5  Endeavor Air      N337PQ         BDL   
5  2023-01-07            6  Endeavor Air      N336PQ         BDL   
6  2023-01-14            6  Endeavor Air      N311PQ         LGA   
7  2023-01-21            6  Endeavor Air      N917XJ         LGA   
8  2023-01-28            6  Endeavor Air      N336PQ         LGA   
9  2023-01-09            1  Endeavor Air      N491PX         LGA   

   Dep_CityName DepTime_label  Dep_Delay  Dep_Delay_Tag Dep_Delay_Type  ...  \
0  Hartford, CT       Morning         -3              0      Low <5min  ...   
1  Hartford, CT       Morning         -5              0      Low <5min  ...   
2  Hartford, CT       Morning         -5              0      Low <5min  ...   
3  Hartford, CT       Morning         -6              0      Low <5min  ...   
4  Hartford, CT       Morning         -1              0      Low <5min  ...   
5  Hartford, CT       Morning        -10              0      Low <5min  ...   
6  New York, NY     Afternoon         -8              0      Low <5min  ...   
7  New York, NY     Afternoon        -10              0      Low <5min  ...   
8  New York, NY     Afternoon         -5              0      Low <5min  ...   
9  New York, NY       Evening         -7              0      Low <5min  ...   

  Flight_Duration       Distance_type  Delay_Carrier Delay_Weather  Delay_NAS  \
0              56  Short Haul >1500Mi              0             0          0   
1              62  Short Haul >1500Mi              0             0          0   
2              49  Short Haul >1500Mi              0             0          0   
3              54  Short Haul >1500Mi              0             0          0   
4              50  Short Haul >1500Mi              0             0          0   
5              62  Short Haul >1500Mi              0             0          0   
6             117  Short Haul >1500Mi              0             0          0   
7             125  Short Haul >1500Mi              0             0          0   
8             130  Short Haul >1500Mi              0             0          0   
9              63  Short Haul >1500Mi              0             0          0   

  Delay_Security  Delay_LastAircraft           Manufacturer  Model  \
0              0                   0  CANADAIR REGIONAL JET    CRJ   
1              0                   0  CANADAIR REGIONAL JET    CRJ   
2              0                   0  CANADAIR REGIONAL JET    CRJ   
3              0                   0  CANADAIR REGIONAL JET    CRJ   
4              0                   0  CANADAIR REGIONAL JET    CRJ   
5              0                   0  CANADAIR REGIONAL JET    CRJ   
6              0                   0  CANADAIR REGIONAL JET    CRJ   
7              0                   0  CANADAIR REGIONAL JET    CRJ   
8              0                   0  CANADAIR REGIONAL JET    CRJ   
9              0                   0  CANADAIR REGIONAL JET    CRJ   

   Aicraft_age  
0           16  
1           16  
2           10  
3           17  
4           10  
5           10  
6           10  
7           16  
8           10  
9            4  

[10 rows x 24 columns]

In [4]:
# AIRPORTS GEOLOCATION
geo_airports.head(10)

IATA_CODE                              AIRPORT           CITY STATE COUNTRY  \
0       ABE  Lehigh Valley International Airport      Allentown    PA     USA   
1       ABI             Abilene Regional Airport        Abilene    TX     USA   
2       ABQ    Albuquerque International Sunport    Albuquerque    NM     USA   
3       ABR            Aberdeen Regional Airport       Aberdeen    SD     USA   
4       ABY   Southwest Georgia Regional Airport         Albany    GA     USA   
5       ACK           Nantucket Memorial Airport      Nantucket    MA     USA   
6       ACT                Waco Regional Airport           Waco    TX     USA   
7       ACV                       Arcata Airport  Arcata/Eureka    CA     USA   
8       ACY  Atlantic City International Airport  Atlantic City    NJ     USA   
9       ADK                         Adak Airport           Adak    AK     USA   

   LATITUDE  LONGITUDE  
0  40.65236  -75.44040  
1  32.41132  -99.68190  
2  35.04022 -106.60919  
3  45.44906  -98.42183  
4  31.53552  -84.19447  
5  41.25305  -70.06018  
6  31.61129  -97.23052  
7  40.97812 -124.10862  
8  39.36420  -74.42310  
9  51.87796 -176.64603

In [5]:
# MAJ US FLIGHT - JANUARY 2024
maj_us_flight.head(10)

FlightDate  Day_Of_Week       Airline Tail_Number Dep_Airport  \
0  2023-01-02            1  Endeavor Air      N605LR         BDL   
1  2023-01-03            2  Endeavor Air      N605LR         BDL   
2  2023-01-04            3  Endeavor Air      N331PQ         BDL   
3  2023-01-05            4  Endeavor Air      N906XJ         BDL   
4  2023-01-06            5  Endeavor Air      N337PQ         BDL   
5  2023-01-07            6  Endeavor Air      N336PQ         BDL   
6  2023-01-14            6  Endeavor Air      N311PQ         LGA   
7  2023-01-21            6  Endeavor Air      N917XJ         LGA   
8  2023-01-28            6  Endeavor Air      N336PQ         LGA   
9  2023-01-09            1  Endeavor Air      N491PX         LGA   

   Dep_CityName DepTime_label  Dep_Delay  Dep_Delay_Tag Dep_Delay_Type  ...  \
0  Hartford, CT       Morning         -3              0      Low <5min  ...   
1  Hartford, CT       Morning         -5              0      Low <5min  ...   
2  Hartford, CT       Morning         -5              0      Low <5min  ...   
3  Hartford, CT       Morning         -6              0      Low <5min  ...   
4  Hartford, CT       Morning         -1              0      Low <5min  ...   
5  Hartford, CT       Morning        -10              0      Low <5min  ...   
6  New York, NY     Afternoon         -8              0      Low <5min  ...   
7  New York, NY     Afternoon        -10              0      Low <5min  ...   
8  New York, NY     Afternoon         -5              0      Low <5min  ...   
9  New York, NY       Evening         -7              0      Low <5min  ...   

  Flight_Duration       Distance_type  Delay_Carrier Delay_Weather  Delay_NAS  \
0              56  Short Haul >1500Mi              0             0          0   
1              62  Short Haul >1500Mi              0             0          0   
2              49  Short Haul >1500Mi              0             0          0   
3              54  Short Haul >1500Mi              0             0          0   
4              50  Short Haul >1500Mi              0             0          0   
5              62  Short Haul >1500Mi              0             0          0   
6             117  Short Haul >1500Mi              0             0          0   
7             125  Short Haul >1500Mi              0             0          0   
8             130  Short Haul >1500Mi              0             0          0   
9              63  Short Haul >1500Mi              0             0          0   

  Delay_Security  Delay_LastAircraft           Manufacturer  Model  \
0              0                   0  CANADAIR REGIONAL JET    CRJ   
1              0                   0  CANADAIR REGIONAL JET    CRJ   
2              0                   0  CANADAIR REGIONAL JET    CRJ   
3              0                   0  CANADAIR REGIONAL JET    CRJ   
4              0                   0  CANADAIR REGIONAL JET    CRJ   
5              0                   0  CANADAIR REGIONAL JET    CRJ   
6              0                   0  CANADAIR REGIONAL JET    CRJ   
7              0                   0  CANADAIR REGIONAL JET    CRJ   
8              0                   0  CANADAIR REGIONAL JET    CRJ   
9              0                   0  CANADAIR REGIONAL JET    CRJ   

   Aicraft_age  
0           16  
1           16  
2           10  
3           17  
4           10  
5           10  
6           10  
7           16  
8           10  
9            4  

[10 rows x 24 columns]

In [6]:
weather_by_airport.head(10)

time  tavg  tmin  tmax  prcp  snow   wdir  wspd    pres airport_id
0  2023-01-01   8.1   2.2  11.7   0.0   0.0  278.0   9.7  1013.8        ABE
1  2023-01-02   5.4   0.0  11.7   0.0   0.0  353.0   3.6  1019.6        ABE
2  2023-01-03   8.4   7.2   9.4  15.2   0.0   50.0   5.0  1013.9        ABE
3  2023-01-04  11.1   6.7  17.2   0.0   0.0  302.0   4.7  1009.8        ABE
4  2023-01-05  12.7   6.7  14.4   7.9   0.0  292.0   7.2  1013.0        ABE
5  2023-01-06   5.8   2.8   7.2   5.8   0.0  308.0   9.0  1016.6        ABE
6  2023-01-07   3.8   1.7   6.7   0.0   0.0  285.0  11.2  1022.2        ABE
7  2023-01-08   2.2  -2.7   3.9   0.0   0.0  346.0   6.8  1024.5        ABE
8  2023-01-09   2.3  -1.0   5.0   0.0   0.0  271.0  11.9  1015.2        ABE
9  2023-01-10   1.9  -1.6   4.4   0.0   0.0  250.0   9.4  1017.4        ABE

In [7]:
cancelled_diverted.head(10)

FlightDate  Day_Of_Week       Airline Tail_Number  Cancelled  Diverted  \
0  2023-01-25            3  Endeavor Air      N691CA        1.0       0.0   
1  2023-01-03            2  Endeavor Air      N937XJ        1.0       0.0   
2  2023-01-03            2  Endeavor Air      N937XJ        1.0       0.0   
3  2023-01-11            3  Endeavor Air      N926XJ        1.0       0.0   
4  2023-01-25            3  Endeavor Air      N335PQ        0.0       1.0   
5  2023-01-11            3  Endeavor Air      N926XJ        1.0       0.0   
6  2023-01-18            3  Endeavor Air      N678CA        1.0       0.0   
7  2023-01-23            1  Endeavor Air      N183GJ        1.0       0.0   
8  2023-01-03            2  Endeavor Air      N330PQ        1.0       0.0   
9  2023-01-16            1  Endeavor Air      N301PQ        1.0       0.0   

  Dep_Airport                Dep_CityName DepTime_label  Dep_Delay  ...  \
0         JFK                New York, NY       Evening        0.0  ...   
1         MSP             Minneapolis, MN       Morning        0.0  ...   
2         PIT              Pittsburgh, PA     Afternoon        0.0  ...   
3         BWI               Baltimore, MD     Afternoon        0.0  ...   
4         BWI               Baltimore, MD     Afternoon        2.0  ...   
5         JFK                New York, NY     Afternoon        0.0  ...   
6         MSP             Minneapolis, MN       Evening        0.0  ...   
7         LGA                New York, NY     Afternoon        0.0  ...   
8         CID  Cedar Rapids/Iowa City, IA       Morning        0.0  ...   
9         BGR                  Bangor, ME         Night        0.0  ...   

                 Arr_CityName Arr_Delay    Arr_Delay_Type Flight_Duration  \
0         Ithaca/Cortland, NY       0.0  No Arrival Delay             0.0   
1              Pittsburgh, PA       0.0  No Arrival Delay             0.0   
2             Minneapolis, MN       0.0  No Arrival Delay             0.0   
3                New York, NY       0.0  No Arrival Delay             0.0   
4                New York, NY       0.0  No Arrival Delay             0.0   
5               Baltimore, MD       0.0  No Arrival Delay             0.0   
6  Cedar Rapids/Iowa City, IA       0.0  No Arrival Delay             0.0   
7              Louisville, KY       0.0  No Arrival Delay             0.0   
8             Minneapolis, MN       0.0  No Arrival Delay             0.0   
9                New York, NY       0.0  No Arrival Delay             0.0   

   Distance_type Delay_Carrier  Delay_Weather Delay_NAS  Delay_Security  \
0     Short Haul           0.0            0.0       0.0             0.0   
1     Short Haul           0.0            0.0       0.0             0.0   
2     Short Haul           0.0            0.0       0.0             0.0   
3     Short Haul           0.0            0.0       0.0             0.0   
4     Short Haul           0.0            0.0       0.0             0.0   
5     Short Haul           0.0            0.0       0.0             0.0   
6     Short Haul           0.0            0.0       0.0             0.0   
7     Short Haul           0.0            0.0       0.0             0.0   
8     Short Haul           0.0            0.0       0.0             0.0   
9     Short Haul           0.0            0.0       0.0             0.0   

   Delay_LastAircraft  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0  
5                 0.0  
6                 0.0  
7                 0.0  
8                 0.0  
9                 0.0  

[10 rows x 23 columns]

### Task 2: Explore Weather and Aircraft Tables
**Goal:** Practice loading multiple related datasets and understanding relationships.

**Instructions (75 XP):**

1. Load the weather data CSV into `weather_df`
2. Load the aircraft specifications CSV into `aircraft_df`
3. For each dataset, display:
   - First 5 rows
   - Shape
   - Column names
4. Identify potential join keys by examining columns across all three tables
5. Check for common columns (airport codes, aircraft IDs, timestamps)

**Expected Outcome:**
- Three separate DataFrames loaded
- Understanding of how tables relate (airports, aircraft types, timestamps)
- Identification of primary/foreign key relationships

**Validation Checks:**
- All three tables loaded
- Common columns identified (e.g., airport_code, aircraft_type, date)
- Relationship types understood (one-to-many, etc.)

---

In [8]:
# CREATE YOUR CODE HERE

### Task 3: Dataset Size and Coverage Analysis
**Goal:** Understand the scope and temporal coverage of your data.

**Instructions (125 XP):**

1. For flights data:
   - Count unique airports (origin and destination)
   - Count unique airlines
   - Count unique aircraft types
   - Count unique dates
2. For weather data:
   - Count unique weather stations/airports
   - Identify date range (min and max dates)
3. For aircraft data:
   - Count unique aircraft models
   - Identify manufacturer distribution using `.value_counts()`
4. Create a summary DataFrame showing:
   - Dataset name
   - Number of rows
   - Number of columns
   - Date range (if applicable)

**Expected Outcome:**
- Comprehensive overview of dataset coverage
- Understanding of temporal span (which months/days covered)
- Knowledge of how many unique entities (airports, aircraft, airlines)

**Validation Checks:**
- Unique counts calculated for all key dimensions
- Date ranges identified
- Summary table created

---

In [9]:
# CREATE YOUR CODE HERE

### **LEVEL 2: Data Cleaning** (300 XP)

---

### Task 4: Detect and Analyze Missing Values
**Goal:** Learn to identify missing data patterns across aviation datasets.

**Instructions (75 XP):**

1. For each dataset (flights, weather, aircraft), use `.isna().sum()` to count missing values per column
2. Calculate the percentage of missing values: `(missing_count / total_rows) * 100`
3. Create a bar plot showing missing values for the flights dataset
4. Identify which columns have more than 10% missing data
5. Document potential reasons for missing values (e.g., delayed flights might have missing arrival times)

**Expected Outcome:**
- Clear understanding of data completeness
- Visual representation of missing data
- Strategic plan for handling missing values

**Validation Checks:**
- Missing value counts for all datasets
- Percentages calculated
- High-missing columns identified
- Visualization created

---

In [10]:
# CREATE YOUR CODE HERE

### Task 5: Handle Missing Values in Flight Delays
**Goal:** Practice different strategies for handling missing data.

**Instructions (100 XP):**

1. Focus on delay-related columns (departure_delay, arrival_delay)
2. Check if missing delays correlate with cancellations (if cancellation column exists)
3. For missing arrival delays:
   - If flight is cancelled → fill with a special value (e.g., -9999) or drop
   - If flight completed → fill with 0 or median delay
4. Create a new column `delay_data_quality` indicating if delays are:
   - 'complete' (no missing values)
   - 'imputed' (filled values)
   - 'cancelled' (cancelled flights)
5. Document how many rows were affected by each treatment

**Expected Outcome:**
- Clean delay columns ready for analysis
- Documentation of data cleaning decisions
- Quality flag for transparency in later analysis

**Validation Checks:**
- Missing delay values addressed
- Quality flag column created
- Cleaning documentation prepared

---


In [11]:
# CREATE YOUR CODE HERE

### Task 6: Fix Data Types and Remove Duplicates
**Goal:** Ensure data types are correct and datasets are deduplicated.

**Instructions (125 XP):**

1. Convert date/time columns to datetime format:
   ```python
   df['column'] = pd.to_datetime(df['column'])
   ```
2. Convert numeric columns stored as strings to appropriate types (int or float)
3. Convert categorical columns (airline, airport codes) to category dtype for memory efficiency
4. Check for duplicate rows:
   - In flights: check for duplicate flight numbers on same date
   - In weather: check for duplicate timestamps at same location
   - In aircraft: check for duplicate aircraft registrations
5. Remove duplicates using `.drop_duplicates()` with appropriate subset columns
6. Report how many duplicates were found and removed from each dataset

**Expected Outcome:**
- All columns have correct data types
- No duplicate records
- Memory-efficient data structures
- Documentation of data quality improvements

**Validation Checks:**
- Date columns are datetime type
- Numeric columns are int/float
- No duplicates remain
- Memory usage optimized

---

In [12]:
# CREATE YOUR CODE HERE


### **LEVEL 3: Time Operations** (200 XP)

---

### Task 7: Create DateTime Index and Extract Temporal Features
**Goal:** Master datetime indexing and feature extraction for time-series analysis.

**Instructions (100 XP):**

1. Set the flight datetime as the index using `.set_index()`
2. Sort the index using `.sort_index()`
3. Extract temporal features into new columns:
   - `year`, `month`, `day`, `hour`, `day_of_week`, `quarter`
4. Create a `season` column based on month:
   - Winter: Dec, Jan, Feb
   - Spring: Mar, Apr, May
   - Summer: Jun, Jul, Aug
   - Fall: Sep, Oct, Nov
5. Create a `time_of_day` column:
   - Early Morning: 00:00-05:59
   - Morning: 06:00-11:59
   - Afternoon: 12:00-17:59
   - Evening: 18:00-23:59

**Expected Outcome:**
- DataFrame with datetime index
- Rich temporal features for grouping and filtering
- Ready for time-based analysis

**Validation Checks:**
- DateTime index set and sorted
- All temporal features created
- Season and time_of_day categories defined correctly

---

In [13]:
# CREATE YOUR CODE HERE

### Task 8: Time-Based Slicing and Filtering
**Goal:** Practice advanced time-based data selection techniques.

**Instructions (100 XP):**

1. Slice flights for a specific month (e.g., July 2023): `df.loc['2023-07']`
2. Extract all flights during summer months (Jun-Aug)
3. Filter flights during morning hours (06:00-11:59)
4. Find all flights on weekends (Saturday and Sunday)
5. Create a subset of "holiday season" flights (Thanksgiving week + Christmas week)
6. For each subset, calculate:
   - Total number of flights
   - Average delay
   - Percentage of delayed flights (delay > 15 minutes)

**Expected Outcome:**
- Mastery of `.loc[]` with datetime index
- Understanding of temporal patterns in aviation data
- Comparative statistics across different time periods

**Validation Checks:**
- Time-based slices created correctly
- Multiple filtering conditions applied
- Statistics calculated for each subset

---

In [14]:
# CREATE YOUR CODE HERE

### **LEVEL 4: Merging & Joining** (450 XP)

---

### Task 9: Inner Join - Flights with Aircraft Specifications
**Goal:** Practice basic inner join to enrich flight data with aircraft details.

**Instructions (100 XP):**

1. Identify the join key: aircraft registration number or aircraft type code
2. Perform an inner join between flights and aircraft specifications:
   ```python
   flights_aircraft = flights.merge(aircraft, on='aircraft_code', how='inner')
   ```
3. Check the shape before and after merge - did you lose any rows?
4. Verify the merge by displaying sample rows showing flight info + aircraft specs
5. Calculate and compare:
   - Average delay by aircraft manufacturer
   - Average delay by aircraft age (if age column exists)
   - Which aircraft type has the most flights?

**Expected Outcome:**
- Combined dataset with flight and aircraft information
- Understanding of which flights matched to aircraft specs
- Initial insights into aircraft performance

**Validation Checks:**
- Merge completed successfully
- Shape comparison documented
- Aircraft-level delay statistics calculated
- No unexpected data loss

---

In [15]:
# CREATE YOUR CODE HERE

### Task 10: Left Join - Keep All Flights Even Without Aircraft Data
**Goal:** Understand when to use left joins to preserve all primary records.

**Instructions (100 XP):**

1. Perform a left join (flights LEFT JOIN aircraft):
   ```python
   flights_all = flights.merge(aircraft, on='aircraft_code', how='left')
   ```
2. Compare row counts between inner join (Task 9) and left join
3. Identify flights with missing aircraft specifications:
   ```python
   missing_aircraft = flights_all[flights_all['aircraft_column'].isna()]
   ```
4. Analyze the missing aircraft data:
   - Which airlines have most flights with missing aircraft info?
   - Are these older or newer flights?
   - What percentage of total flights lack aircraft data?
5. Create a `has_aircraft_data` boolean column

**Expected Outcome:**
- Complete flight dataset with optional aircraft details
- Understanding of data coverage gaps
- Strategic decisions about handling missing joins

**Validation Checks:**
- Left join preserves all flights
- Missing aircraft flights identified and analyzed
- Data quality flag created

---

In [16]:
# CREATE YOUR CODE HERE


### Task 11: Time-Aligned Merge - Flights with Nearest Weather Data
**Goal:** Master `merge_asof()` for matching time-series data by nearest timestamp.

**Instructions (125 XP):**

1. Prepare both datasets:
   - Convert flight departure time to datetime
   - Convert weather timestamp to datetime
   - Sort both by their time columns
2. Use `merge_asof()` to match each flight with the nearest **previous** weather reading:
   ```python
   flights_weather = pd.merge_asof(
       flights_sorted,
       weather_sorted,
       left_on='departure_time',
       right_on='weather_time',
       by='airport_code',  # Match within same airport
       direction='backward'
   )
   ```
3. Verify the merge quality:
   - Check time difference between flight and matched weather reading
   - Ensure no flight is matched with future weather data
4. Create a new column `weather_recency` showing minutes between weather reading and flight

**Expected Outcome:**
- Flights matched with contextual weather at departure
- Time-aligned dataset ready for weather impact analysis
- No temporal data leakage (future weather predicting past flights)

**Validation Checks:**
- All flights have matched weather data
- Weather timestamps are before or equal to flight times
- Time differences are reasonable (< 60 minutes)
- Airport-specific matches confirmed

---

In [17]:
# CREATE YOUR CODE HERE

### Task 12: Multi-Table Merge - Complete Aviation Dataset
**Goal:** Chain multiple merges to create comprehensive analytical dataset.

**Instructions (125 XP):**

1. Start with flights as the base table
2. Merge with aircraft specifications (left join)
3. Merge result with weather data using `merge_asof()` for departure airport
4. If you have airport metadata, merge that too using airport codes
5. Validate the final dataset:
   - Check for unexpected row multiplication
   - Verify all merges preserved intended relationships
   - Use `.shape` at each step to track row counts
6. Create a data dictionary documenting:
   - Original tables used
   - Join keys
   - Join types
   - Final column count and meaning

**Expected Outcome:**
- Single comprehensive dataset with flights, aircraft, weather, and airport info
- Complete context for each flight
- Ready for advanced analysis

**Validation Checks:**
- All merges completed in correct order
- Row counts make sense (no unexpected explosion)
- Key columns from all tables present
- Data dictionary created

---

In [18]:
# CREATE YOUR CODE HERE

### **LEVEL 5: Feature Engineering** (200 XP)

---

### Task 13: Engineer Delay Categories and Performance Metrics
**Goal:** Create meaningful categorical features from continuous delay data.

**Instructions (100 XP):**

1. Create delay severity categories:
   ```python
   def categorize_delay(minutes):
       if minutes <= 0: return 'On Time/Early'
       elif minutes <= 15: return 'Slight Delay'
       elif minutes <= 45: return 'Moderate Delay'
       elif minutes <= 120: return 'Significant Delay'
       else: return 'Severe Delay'
   ```
2. Apply to both departure and arrival delays
3. Create a binary column `is_delayed` (True if delay > 15 minutes)
4. Calculate `delay_worsened` (arrival delay > departure delay)
5. Create performance score: `100 - (delay_minutes / 10)` capped at 0-100
6. Compute delay propagation: `arrival_delay - departure_delay`

**Expected Outcome:**
- Rich categorical features for grouping
- Binary flags for filtering
- Continuous performance metrics
- Understanding of delay progression during flight

**Validation Checks:**
- All categories defined correctly
- Binary flags work as expected
- Performance scores calculated
- No null values in new columns

---


In [19]:
# CREATE YOUR CODE HERE


### Task 14: Weather Impact Features
**Goal:** Engineer features capturing weather's effect on flight operations.

**Instructions (100 XP):**

1. Create adverse weather flags:
   - `heavy_wind` (wind speed > 20 mph)
   - `low_visibility` (visibility < 3 miles)
   - `precipitation` (rain or snow present)
   - `extreme_temp` (temp < 32°F or > 95°F)
2. Create composite `adverse_weather_score`:
   - Count how many adverse conditions are present (0-4)
3. Categorize weather conditions:
   - Ideal: score = 0
   - Fair: score = 1
   - Poor: score = 2
   - Severe: score >= 3
4. For flights with weather data, calculate:
   - Correlation between adverse weather and delays
   - Average delay by weather category

**Expected Outcome:**
- Quantified weather impact on operations
- Easy-to-interpret weather severity scale
- Foundation for predictive modeling

**Validation Checks:**
- Weather flags created correctly
- Composite score calculated
- Weather categories assigned
- Correlation analysis completed

---

In [20]:
# CREATE YOUR CODE HERE

### **LEVEL 6: Aggregation & Analysis** (300 XP)

---

### Task 15: Aircraft Type Performance Analysis
**Goal:** Use groupby operations to compare aircraft performance.

**Instructions (100 XP):**

1. Group flights by aircraft type and calculate:
   - Mean departure delay
   - Mean arrival delay
   - Percentage of flights delayed (>15 min)
   - Total number of flights
   - Standard deviation of delays
2. Sort aircraft by average delay (worst to best)
3. Create a bar plot showing top 10 aircraft with highest average delays
4. Filter for aircraft types with at least 100 flights (statistical significance)
5. Calculate reliability score: `100 - (pct_delayed * mean_delay / 10)`

**Expected Outcome:**
- Ranking of aircraft by operational performance
- Statistical validity through minimum sample sizes
- Visual comparison of aircraft types
- Actionable insights for fleet management

**Validation Checks:**
- Groupby aggregations calculated
- Sample size filtering applied
- Visualization created
- Reliability scores computed

---

In [21]:
# CREATE YOUR CODE HERE

### Task 16: Airline and Airport Efficiency Comparison
**Goal:** Compare operational efficiency across airlines and airports.

**Instructions (100 XP):**

1. **Airline Analysis:**
   - Group by airline, calculate mean delays, on-time percentage, total flights
   - Rank airlines by on-time performance
   - Identify which airlines have most consistent performance (lowest std dev)

2. **Airport Analysis:**
   - Group by origin airport, calculate departure delay statistics
   - Group by destination airport, calculate arrival delay statistics
   - Identify "best" airports (lowest delays, highest on-time %)
   - Find "challenging" airports (highest delays)

3. Create pivot table:
   - Rows: Airlines
   - Columns: Top 10 airports
   - Values: Average delay
   - This shows which airline-airport combinations perform best/worst

**Expected Outcome:**
- Comprehensive airline performance rankings
- Airport-specific operational insights
- Cross-tabulated analysis revealing specific problematic routes
- Data-driven recommendations for travelers

**Validation Checks:**
- Both airline and airport analyses complete
- Rankings created
- Pivot table generated
- Best/worst performers identified

---

In [22]:
# CREATE YOUR CODE HERE

### Task 17: Temporal Patterns - Seasonality and Time-of-Day Analysis
**Goal:** Uncover temporal patterns in flight delays using groupby and pivot tables.

**Instructions (100 XP):**

1. **Seasonal Analysis:**
   - Group by season, calculate average delays
   - Create line plot showing delay trends across months
   - Identify peak delay season

2. **Time-of-Day Analysis:**
   - Group by hour of day, calculate average delays
   - Plot delay pattern across 24 hours
   - Identify worst time slots for delays

3. **Day-of-Week Analysis:**
   - Group by day of week, calculate statistics
   - Determine if weekends differ from weekdays

4. **Create heatmap pivot table:**
   - Rows: Day of week
   - Columns: Hour of day
   - Values: Average delay
   - This reveals the worst day-hour combinations

**Expected Outcome:**
- Understanding of when delays are most likely
- Visual representations of temporal patterns
- Strategic insights for flight booking

**Validation Checks:**
- All temporal dimensions analyzed
- Visualizations created
- Heatmap shows clear patterns
- Peak delay periods identified

---

In [23]:
# CREATE YOUR CODE HERE

### **LEVEL 7: Advanced Analytics** (400 XP)

---

### Task 18: Weather Impact Quantification
**Goal:** Perform advanced analysis of weather's causal effect on delays.

**Instructions (125 XP):**

1. **Comparative Analysis:**
   - Calculate average delay in ideal weather conditions
   - Calculate average delay in adverse weather conditions
   - Compute the weather impact: `delay_adverse - delay_ideal`

2. **Correlation Analysis:**
   - Calculate correlation between each weather variable and delay:
     - Temperature vs delay
     - Wind speed vs delay
     - Visibility vs delay
     - Precipitation vs delay
   - Create scatter plots for strongest correlations

3. **Conditional Probability:**
   - P(delay | adverse weather)
   - P(delay | ideal weather)
   - Calculate relative risk: how many times more likely is delay in bad weather?

4. **Multi-factor Analysis:**
   - Create categories combining weather + time-of-day
   - Find worst combinations (e.g., "heavy rain + morning rush")

**Expected Outcome:**
- Quantified weather impact on operations
- Statistical evidence of weather's role
- Identification of high-risk weather conditions
- Data-driven weather threshold recommendations

**Validation Checks:**
- Comparative statistics calculated
- Correlations computed and visualized
- Conditional probabilities determined
- Multi-factor insights derived

---

In [24]:
# CREATE YOUR CODE HERE

### Task 19: Route-Level Performance Analysis
**Goal:** Analyze origin-destination pairs to identify problematic routes.

**Instructions (125 XP):**

1. Create `route` column combining origin and destination: `origin + '-' + destination`
2. Group by route and calculate:
   - Total flights
   - Average delay
   - On-time percentage
   - Most common aircraft types used
   - Average weather score
3. Filter for routes with at least 50 flights
4. Identify:
   - Top 10 most reliable routes (best on-time %)
   - Top 10 most problematic routes (worst delays)
   - Routes where delays worsen during flight (arrival > departure)
5. For worst routes, analyze contributing factors:
   - Is it weather-related?
   - Is it aircraft-related?
   - Is it specific to certain airlines?
   - Is it time-of-day related?
6. Create visualization showing geographic distribution of route performance (if coordinates available)

**Expected Outcome:**
- Route-specific performance rankings
- Root cause analysis for problem routes
- Recommendations for route optimization
- Strategic insights for airline operations

**Validation Checks:**
- Routes created and analyzed
- Filtering by sample size applied
- Best/worst routes identified
- Contributing factor analysis completed
- Actionable recommendations developed

---



In [25]:
# CREATE YOUR CODE HERE

### Task 20: Comprehensive Aviation Dashboard Summary
**Goal:** Synthesize all analyses into a comprehensive aviation performance report.

**Instructions (150 XP):**

1. **Executive Summary Section:**
   - Overall fleet statistics (total flights, airports, aircraft)
   - System-wide on-time performance
   - Total delay minutes and economic impact estimate
   - Key trends (improving or worsening over time)

2. **Top Performers Section:**
   - Best airline (by on-time %)
   - Most reliable aircraft type
   - Best-performing airports
   - Most reliable routes

3. **Areas for Improvement:**
   - Worst airline performance
   - Most problematic aircraft
   - Airports with highest delays
   - Routes requiring attention

4. **Root Cause Analysis:**
   - Primary delay factors ranked:
     - Weather contribution
     - Aircraft age/type contribution  
     - Airport congestion patterns
     - Time-of-day effects
   - Quantify each factor's impact

5. **Predictive Insights:**
   - Create delay risk score for any flight based on:
     - Route
     - Aircraft type
     - Weather forecast
     - Time of day
     - Day of week
   - Test on sample flights

6. **Visualizations:**
   - Overall delay distribution histogram
   - Monthly trend line
   - Weather impact comparison chart
   - Top 10 aircraft/airlines comparison
   - Geographic delay heatmap (if possible)

7. **Recommendations:**
   - Data-driven suggestions for:
     - Fleet optimization
     - Schedule adjustments
     - Weather contingency planning
     - Route modifications

**Expected Outcome:**
- Professional-grade analytical report
- Clear visualizations communicating insights
- Actionable business recommendations
- Demonstration of end-to-end data analysis skills
- Portfolio-ready project deliverable

**Validation Checks:**
- All major findings summarized
- Multiple visualization types created
- Statistical rigor maintained
- Business recommendations provided
- Report is clear and actionable
- Code is well-documented and reproducible

---



In [26]:
# CREATE YOUR CODE HERE